source/dub/generators/build.d

```diff
@@ -552,6 +553,99 @@ class BuildGenerator : ProjectGenerator {
 	}
 }
 
+
+/** Runs a list of build commands for a particular package and scan for "dub:" lines output
+
+	This function sets all DUB speficic environment variables and scan for
+	"dub:" lines output and amend build_settings with the settings passed from
+	the commands through these lines.
+
+	Each command get its environment updated by the settings of the previous commands
+*/
+private void runScanBuildCommands(in string[] commands, in Package pack, in Project proj,
+	in GeneratorSettings settings, ref BuildSettings build_settings)
+{
+	import std.process : Config, pipe, spawnShell, wait;
+	import std.stdio : File, stdin, stderr, stdout;
+
+	auto env = prepareCommandsEnvironment(pack, proj, settings, build_settings);
+
+	version(Windows) enum nullFile = "NUL";
+	else version(Posix) enum nullFile = "/dev/null";
+	else static assert(0);
+
+	// Do not use retainStdout here as the output reading loop would hang forever.
+	// Config.retainStderr is not necessary with stderr.
+	const config = Config.none;
+	const logLevel = getLogLevel();
+
+	auto childrenErr = logLevel >= LogLevel.none ? File(nullFile, "w") : stderr;
+
+	foreach (cmd; commands) {
+		logDiagnostic("Running %s", cmd);
+		auto p = pipe();
+		auto pid = spawnShell(cmd, stdin, p.writeEnd, childrenErr, env, config);
+
+		string[] dubLines;
+
+		foreach (l; p.readEnd.byLine) {
+			if (l.startsWith("dub:")) {
```

Comment:
Not sure what you mean exactly nor how to do it (I'm not too familiar yet with all parts of the code).
Can this work with the SDL packages?

Reply:
OK got it.
So your suggestion is to output JSON text directly from the scripts?
I need to delimitate it somehow as scripts need to be able to log things (and often we don't have control over it - think CMake), so we should only allow one liners JSON strings (still with `"dub:"` prefix).
Each script can echo several disjoint JSON strings, that Dub merges behind scene.

Reply:
yeah I would probably always go for single line json strings. Maybe it would actually be a better idea not allowing all these settings though, so actually I think how it is right now would be enough for now. There is a problem though: what is `dub describe` going to output? Will it run the preBuildScript or only go from what is in the dub.json?

Reply:
I did 853ee11a9bd6e22067c9af98943e22000828c522 which keeps control over what is parsed in JSON. It works but makes scripting more complex than necessary imo. The flexibility won is only faked and I think I should revert that.

Didn't look into `dub describe` yet. My assumption is that it issues `preGenerateCommands`, but not `preBuildCommands`.